ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -48,32 +53,130 @@ private class PackageService(reloadPackageStoreIfChanged: PackageService.ReloadP
   @volatile private var state: State =
     State(Set.empty, TemplateIdMap.Empty, Map.empty, Map.empty, Map.empty)
 
-  // synchronized, so two threads cannot reload it concurrently
-  def reload(implicit
-      ec: ExecutionContext,
-      lc: LoggingContextOf[InstanceUUID],
-  ): Future[Error \/ Unit] =
-    synchronized {
-      reloadPackageStoreIfChanged(state.packageIds).map {
-        _.map {
-          case Some(diff) =>
-            this.state = this.state.append(diff)
-            logger.info(s"new package IDs loaded: ${diff.keySet.mkString(", ")}")
-            logger.debug(s"loaded diff: $diff")
-            ()
-          case None =>
-            logger.debug(s"new package IDs not found")
-            ()
+  // volatile, because we want to reduce the probability of unnecessary package updates
+  // due to multiple threads reading this.
+  // Writes to this are synchronized with a mutex, so in theory this should work out fine.
+  @volatile private var lastUpdated = LocalDateTime.MIN
+
+  // Regular updates should happen regardless of the current state every minute.
+  private def packagesShouldBeFetchedAgain =
+    lastUpdated.until(LocalDateTime.now(), temporal.ChronoUnit.SECONDS) >= 60L
+
+  private val _reload = {
+    // Because things can go so easily wrong, the declaration of the mutex
+    // was moved into a separate block, so it's easier to track down all interactions.
+    val packageUpdateSync = util.Concurrent.Mutex()
+    val updateTaskUpdateSync = util.Concurrent.Mutex()
+    @volatile var ongoingPackageUpdateTask = Future.successful(().right)
+    @volatile var isValid = false
+
+    def reload(jwt: Jwt)(implicit
+        ec: ExecutionContext,
+        lc: LoggingContextOf[InstanceUUID],
+    ): Future[Error \/ Unit] = {
+      def update() = {
+        val f =
+          EitherT
+            .eitherT(
+              Future(
+                logger.debug("Trying to execute a package update")
```

Comment:
can we get rid of this `Future` to do debug logging ? would the below not work ?
```scala
EitherT.eitherT({
                logger.debug("Trying to execute a package update")
                reloadPackageStoreIfChanged(jwt)(state.packageIds)
            })
```

Reply:
To my knowledge `EitherT.eitherT` is not lazily evaluating it's parameter.

Reply:
however I think the result of using your code would in this case be the same

Reply:
ahh ok cheers, my suggestion was motivated by an effort to reduce object creation where possible as long as we don't loose functionality or readability.